10_rest/n0npax/cmd/puppy-server/main_test.go

```diff
@@ -0,0 +1,166 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"os"
+	"testing"
+
+	"bou.ke/monkey"
+
+	puppy "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy"
+	store "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy/store"
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseArgs(t *testing.T) {
+
+	os.Args = []string{"", "-s", "map", "-d", "/dev/null", "--port", "1234"}
+	config, err := parseArgs()
+	assert.NoError(t, err)
+	assert.Equal(t, config.port, 1234)
+	assert.Equal(t, config.path, "/dev/null")
+	assert.Equal(t, config.sType, "map")
+}
+
+func TestMainWithFakeArgs(t *testing.T) {
+	fakeExit := func(int) {
+		panic("foo-arg")
+	}
+	patch := monkey.Patch(os.Exit, fakeExit)
+	defer patch.Unpatch()
+
+	os.Args = []string{"", "-s", "map", "--foo"}
+	assert.PanicsWithValuef(t, "foo-arg", main, "foo-arg")
+}
+
+func TestMain(t *testing.T) {
+	fakeRunSrv := func(*gin.Engine, int) error {
+		panic("main")
+	}
+	patch := monkey.Patch(runSrv, fakeRunSrv)
+	defer patch.Unpatch()
+
+	os.Args = []string{"", "-s", "map", "-d", "../../test-data/data.json"}
+	assert.PanicsWithValuef(t, "main", main, "main")
```

Comment:
This one is tricky. However I think with a dodgy port you can still get 100% coverage and circumvent monkey patching here.

Reply:
passing invalid port `-22` or `678678` also calls `os.Exit`